### ISSUE TITLE
`useHookAtTopLevel` reports errors on code that is not React

### ISSUE THREAD

**Issue by ssilve1989** [biomejs/biome#3476]
### Environment information

```block
CLI:
  Version:                      1.8.3
  Color support:                true

Platform:
  CPU Architecture:             aarch64
  OS:                           macos

Environment:
  BIOME_LOG_DIR:                unset
  NO_COLOR:                     unset
  TERM:                         "xterm-256color"
  JS_RUNTIME_VERSION:           "v20.11.1"
  JS_RUNTIME_NAME:              "node"
  NODE_PACKAGE_MANAGER:         "pnpm/8.10.2"

Biome Configuration:
  Status:                       Loaded successfully
  Formatter disabled:           false
  Linter disabled:              false
  Organize imports disabled:    false
  VCS disabled:                 false

Workspace:
  Open Documents:               0
```

### What happened?

The following sample NestJS snippet triggers the `useHookAtTopLevel` thinking `useGlobalInterceptors` is a React hook. 
```typescript
async function bootstrap() {
  const app = await NestFactory.create(AppModule)
  app.useGlobalInterceptors(new RandomInterceptor());
}

bootstrap();
```

### Expected result

It should not raise any warning/error

### Code of Conduct

- [X] I agree to follow Biome's Code of Conduct

**Comment by yossydev** [comment]
Reproduced here. It is indeed an error.
https://github.com/yossydev/use-hook-at-top-level-nextjs

I'm not familiar with nestjs and don't know what use you are using useHookAtTopLevel for, but I would have thought it would be better not to check lint as this is a React rule in the first place.

**Comment by dyc3** [comment]
That link 404s.

This rule _should_ only be checked in `jsx` or `tsx` environments. I don't think we currently have a way to automatically determine if code is react code or not other than that.

**Comment by yossydev** [comment]
> That link 404s.

Sorry, this was private repository.
you can see it now.

> This rule should only be checked in jsx or tsx environments. I don't think we currently have a way to automatically determine if code is react code or not other than that.

Certainly, that sounds better.

**Comment by Javimtib92** [comment]
I think we could use a couple of heuristics like checking is_react_component (checks if function starts with Uppercase like react components do) and checking if the file is `jsx`/`tsx`.

Could I take this task to try that out?

**Comment by dyc3** [comment]
Sure! I'll go ahead and assign you.

**Comment by arendjr** [comment]
Please be aware we already have such a heuristic: https://github.com/biomejs/biome/blob/d3dacc995501f3bc7c9195390b49b921b2b5f14c/crates/biome_js_analyze/src/react/hooks.rs#L132

I suppose this is not used by the rule yet…

**Comment by Javimtib92** [comment]
> Please be aware we already have such a heuristic:
> 
> https://github.com/biomejs/biome/blob/d3dacc995501f3bc7c9195390b49b921b2b5f14c/crates/biome_js_analyze/src/react/hooks.rs#L132
> 
> I suppose this is not used by the rule yet…

Yep, that's the function I wanted to use. I was thinking now that maybe limiting this to `jsx` or `tsx` files is not a good heuristic because a hook could be used inside another hook defined in a `js` or `ts` file, right?

For example: https://github.com/ilyalesik/react-media-hook/blob/master/index.js

**Comment by Javimtib92** [comment]
Investigating this I noticed that eslint approach is not perfect either:

<img width="1183" alt="image" src="https://github.com/user-attachments/assets/9e5a14d9-fc3e-4646-b589-e8ab14afce27">

**Comment by Javimtib92** [comment]
A similar case was fixed for some jest functions that start with “use”

https://github.com/biomejs/biome/blob/d3dacc995501f3bc7c9195390b49b921b2b5f14c/crates/biome_js_analyze/src/react/hooks.rs#L156

I tried adding other heuristics, like checking that the root caller is a react component or a hook, but so far all I tried ends up changing the snapshots because some cases get uncovered. We might need to do some comprimises with this.

**Comment by dyc3** [comment]
It might be useful if biome could inspect `package.json` to guess what environment the file is in.

**Comment by Javimtib92** [comment]
@dyc3 I think your suggestion is actually a good workaround. I opened a Draft PR with this change. I need to find a way to run the snapshots simulating an environment with the react dependency.  I would highly appreciate any suggestions. Thanks!

**Comment by Conaclos** [comment]
There is something I don't understand: the rule is not enabled by default (it is not recommended). Thus, the rule should only be enabled if the project use React. Why did you enable it in the first place?

**Comment by ematipico** [comment]
Maybe the documentation could be improved, but this rule isn't meant for non-react projects, so I don't think it's an issue on our side.

**Comment by dyc3** [comment]
I agree with @ematipico  I'm all for saving user's time configuring projects, and maybe we could come up with some really good heuristics to auto detect what kind of project biome is operating on, but heuristics are not perfect. If anything, it would probably be better to ask the user for that info in `biome init` when they are setting it up.

Would it make sense to have more collections of rules similar to how we currently have recommended rules? eg. `react-recommended`, `react-all`. Or we could give rules tags to do something similar, though tags might be a bit more complicated on the implementation side.

**Comment by ematipico** [comment]
> Would it make sense to have more collections of rules similar to how we currently have recommended rules?

There's an RFC for that, you can find it from the pinned umbrella issue "Biome developments"

**Comment by ssilve1989** [comment]
This has now returned in Biome v2.0.5. The rule is not enabled as per the rage output after running `biome migrate`, but it is erroring again on Nest files as in the initial issue. 

This seems related to Domains, since that is a recommended rule in the React domain. https://biomejs.dev/linter/javascript/rules/

but in a project that includes both Nest and React in its dependencies its going to auto-activate this React rule, but apply it to Nest files. I guess the solution here is to use overrides somehow and avoid having it run on the Nest files

```
CLI:
  Version:                      2.0.5
  Color support:                true

Platform:
  CPU Architecture:             aarch64
  OS:                           macos

Environment:
  BIOME_LOG_PATH:               unset
  BIOME_LOG_PREFIX_NAME:        unset
  BIOME_CONFIG_PATH:            unset
  BIOME_THREADS:                unset
  NO_COLOR:                     unset
  TERM:                         xterm-256color
  JS_RUNTIME_VERSION:           v22.14.0
  JS_RUNTIME_NAME:              node
  NODE_PACKAGE_MANAGER:         pnpm/9.7.0

Biome Configuration:
  Status:                       Loaded successfully
  Path:                         biome.json
  Formatter enabled:            true
  Linter enabled:               true
  Assist enabled:               true
  VCS enabled:                  true

Linter:
  JavaScript enabled:           unset
  JSON enabled:                 unset
  CSS enabled:                  unset
  GraphQL enabled:              unset
  Recommended:                  true
  Enabled rules:
    a11y/noAccessKey
    a11y/noAriaHiddenOnFocusable
    a11y/noAriaUnsupportedElements
    a11y/noAutofocus
    a11y/noDistractingElements
    a11y/noHeaderScope
    a11y/noInteractiveElementToNoninteractiveRole
    a11y/noNoninteractiveElementToInteractiveRole
    a11y/noNoninteractiveTabindex
    a11y/noPositiveTabindex
    a11y/noRedundantAlt
    a11y/noRedundantRoles
    a11y/noStaticElementInteractions
    a11y/noSvgWithoutTitle
    a11y/useAltText
    a11y/useAnchorContent
    a11y/useAriaActivedescendantWithTabindex
    a11y/useAriaPropsForRole
    a11y/useAriaPropsSupportedByRole
    a11y/useButtonType
    a11y/useFocusableInteractive
    a11y/useGenericFontNames
    a11y/useHeadingContent
    a11y/useHtmlLang
    a11y/useIframeTitle
    a11y/useKeyWithMouseEvents
    a11y/useMediaCaption
    a11y/useSemanticElements
    a11y/useValidAnchor
    a11y/useValidAriaProps
    a11y/useValidAriaRole
    a11y/useValidAriaValues
    a11y/useValidAutocomplete
    a11y/useValidLang
    complexity/noAdjacentSpacesInRegex
    complexity/noArguments
    complexity/noBannedTypes
    complexity/noCommaOperator
    complexity/noEmptyTypeParameters
    complexity/noExcessiveCognitiveComplexity
    complexity/noExtraBooleanCast
    complexity/noFlatMapIdentity
    complexity/noForEach
    complexity/noStaticOnlyClass
    complexity/noThisInStatic
    complexity/noUselessCatch
    complexity/noUselessConstructor
    complexity/noUselessContinue
    complexity/noUselessEmptyExport
    complexity/noUselessEscapeInRegex
    complexity/noUselessFragments
    complexity/noUselessLabel
    complexity/noUselessLoneBlockStatements
    complexity/noUselessRename
    complexity/noUselessStringRaw
    complexity/noUselessSwitchCase
    complexity/noUselessTernary
    complexity/noUselessThisAlias
    complexity/noUselessTypeConstraint
    complexity/noUselessUndefinedInitialization
    complexity/useArrowFunction
    complexity/useDateNow
    complexity/useFlatMap
    complexity/useLiteralKeys
    complexity/useNumericLiterals
    complexity/useOptionalChain
    complexity/useRegexLiterals
    complexity/useSimpleNumberKeys
    complexity/useSimplifiedLogicExpression
    correctness/noChildrenProp
    correctness/noConstAssign
    correctness/noConstantCondition
    correctness/noConstantMathMinMaxClamp
    correctness/noConstructorReturn
    correctness/noEmptyCharacterClassInRegex
    correctness/noEmptyPattern
    correctness/noGlobalObjectCalls
    correctness/noInnerDeclarations
    correctness/noInvalidBuiltinInstantiation
    correctness/noInvalidConstructorSuper
    correctness/noInvalidDirectionInLinearGradient
    correctness/noInvalidGridAreas
    correctness/noInvalidPositionAtImportRule
    correctness/noInvalidUseBeforeDeclaration
    correctness/noMissingVarFunction
    correctness/noNonoctalDecimalEscape
    correctness/noPrecisionLoss
    correctness/noSelfAssign
    correctness/noSetterReturn
    correctness/noStringCaseMismatch
    correctness/noSwitchDeclarations
    correctness/noUnknownFunction
    correctness/noUnknownMediaFeatureName
    correctness/noUnknownProperty
    correctness/noUnknownPseudoClass
    correctness/noUnknownPseudoElement
    correctness/noUnknownTypeSelector
    correctness/noUnknownUnit
    correctness/noUnmatchableAnbSelector
    correctness/noUnreachable
    correctness/noUnreachableSuper
    correctness/noUnsafeFinally
    correctness/noUnsafeOptionalChaining
    correctness/noUnusedFunctionParameters
    correctness/noUnusedImports
    correctness/noUnusedLabels
    correctness/noUnusedPrivateClassMembers
    correctness/noUnusedVariables
    correctness/noVoidElementsWithChildren
    correctness/noVoidTypeReturn
    correctness/useIsNan
    correctness/useValidForDirection
    correctness/useValidTypeof
    correctness/useYield
    nursery/useUniqueElementIds
    performance/noAccumulatingSpread
    performance/noDynamicNamespaceImportAccess
    security/noBlankTarget
    security/noDangerouslySetInnerHtml
    security/noDangerouslySetInnerHtmlWithChildren
    security/noGlobalEval
    style/noDescendingSpecificity
    style/noEnum
    style/noInferrableTypes
    style/noNonNullAssertion
    style/noParameterAssign
    style/noUnusedTemplateLiteral
    style/useArrayLiterals
    style/useAsConstAssertion
    style/useConst
    style/useDefaultParameterLast
    style/useDeprecatedReason
    style/useEnumInitializers
    style/useExponentiationOperator
    style/useExportType
    style/useLiteralEnumMembers
    style/useNodejsImportProtocol
    style/useNumberNamespace
    style/useSelfClosingElements
    style/useShorthandFunctionType
    style/useTemplate
    suspicious/noApproximativeNumericConstant
    suspicious/noArrayIndexKey
    suspicious/noAssignInExpressions
    suspicious/noAsyncPromiseExecutor
    suspicious/noCatchAssign
    suspicious/noClassAssign
    suspicious/noCommentText
    suspicious/noCompareNegZero
    suspicious/noConfusingLabels
    suspicious/noConfusingVoidType
    suspicious/noConsole
    suspicious/noConstEnum
    suspicious/noControlCharactersInRegex
    suspicious/noDebugger
    suspicious/noDocumentCookie
    suspicious/noDoubleEquals
    suspicious/noDuplicateAtImportRules
    suspicious/noDuplicateCase
    suspicious/noDuplicateClassMembers
    suspicious/noDuplicateCustomProperties
    suspicious/noDuplicateElseIf
    suspicious/noDuplicateFields
    suspicious/noDuplicateFontNames
    suspicious/noDuplicateJsxProps
    suspicious/noDuplicateObjectKeys
    suspicious/noDuplicateParameters
    suspicious/noDuplicateProperties
    suspicious/noDuplicateSelectorsKeyframeBlock
    suspicious/noEmptyBlock
    suspicious/noEmptyInterface
    suspicious/noExplicitAny
    suspicious/noExtraNonNullAssertion
    suspicious/noFallthroughSwitchClause
    suspicious/noFunctionAssign
    suspicious/noGlobalAssign
    suspicious/noGlobalIsFinite
    suspicious/noGlobalIsNan
    suspicious/noImplicitAnyLet
    suspicious/noImportAssign
    suspicious/noImportantInKeyframe
    suspicious/noIrregularWhitespace
    suspicious/noLabelVar
    suspicious/noMisleadingCharacterClass
    suspicious/noMisleadingInstantiator
    suspicious/noMisrefactoredShorthandAssign
    suspicious/noOctalEscape
    suspicious/noPrototypeBuiltins
    suspicious/noRedeclare
    suspicious/noRedundantUseStrict
    suspicious/noSelfCompare
    suspicious/noShadowRestrictedNames
    suspicious/noShorthandPropertyOverrides
    suspicious/noSparseArray
    suspicious/noSuspiciousSemicolonInJsx
    suspicious/noTemplateCurlyInString
    suspicious/noThenProperty
    suspicious/noUnsafeDeclarationMerging
    suspicious/noUnsafeNegation
    suspicious/noWith
    suspicious/useAdjacentOverloadSignatures
    suspicious/useDefaultSwitchClauseLast
    suspicious/useGetterReturn
    suspicious/useGoogleFontDisplay
    suspicious/useIsArray
    suspicious/useNamespaceKeyword

Workspace:
  Open Documents:               0
```
